File: backend/node_server/trafficMonitor.ts
import { logToSystem } from './logger.ts';

interface Packet {
  timestamp: string;
  source: string;
  destination: string;
  protocol: 'TCP' | 'UDP' | 'HTTPS' | 'DNS';
  size: number;
  threatLevel: 'low' | 'medium' | 'high';
}

let activeStreams: Packet[] = [];
let ioInstance: any = null;

export function setTrafficIo(io: any) {
  ioInstance = io;
}

const COMMON_IPS = [
  '192.168.1.5',
  '192.168.1.12',
  '192.168.1.100',
  '1.1.1.1',
  '8.8.8.8',
  '142.250.180.14', // Google
  '31.13.72.36',    // Facebook
  '52.95.120.50'     // AWS
];

const PROTOCOLS: ('TCP' | 'UDP' | 'HTTPS' | 'DNS')[] = ['TCP', 'UDP', 'HTTPS', 'DNS'];

export function generateLiveTraffic() {
  const packet: Packet = {
    timestamp: new Date().toLocaleTimeString(),
    source: COMMON_IPS[Math.floor(Math.random() * 3)], // Internal
    destination: COMMON_IPS[3 + Math.floor(Math.random() * 5)], // External
    protocol: PROTOCOLS[Math.floor(Math.random() * PROTOCOLS.length)],
    size: Math.floor(Math.random() * 1500),
    threatLevel: Math.random() > 0.98 ? 'high' : (Math.random() > 0.9 ? 'medium' : 'low')
  };

  // Keep last 50
  activeStreams.unshift(packet);
  if (activeStreams.length > 50) activeStreams.pop();

  if (ioInstance) {
    ioInstance.emit('traffic:packet', packet);
  }

  if (packet.threatLevel === 'high') {
    logToSystem('WARN', `FIREWALL ALERT: Malicious ${packet.protocol} pattern detected from ${packet.source} -> ${packet.destination}`);
  }
}

export function startTrafficMonitor() {
  logToSystem('INFO', 'Starting BPF Packet Interceptor service...');
  setInterval(generateLiveTraffic, 800);
}

export function getTrafficHistory() {
  return activeStreams;
}

File: backend/python_server/traffic_monitor.py
import random
import time
from .logger import log_to_system

active_streams = []

def generate_live_traffic():
    packet = {
        "timestamp": time.strftime("%H:%M:%S"),
        "source": f"192.168.1.{random.randint(2, 20)}",
        "destination": f"8.8.8.{random.randint(1, 8)}",
        "protocol": random.choice(['TCP', 'UDP', 'HTTPS', 'DNS']),
        "size": random.randint(40, 1500),
        "threatLevel": 'high' if random.random() > 0.98 else ('medium' if random.random() > 0.9 else 'low')
    }

    active_streams.insert(0, packet)
    if len(active_streams) > 50:
        active_streams.pop()

    if packet['threatLevel'] == 'high':
        log_to_system('WARN', f"FIREWALL ALERT: Malicious {packet['protocol']} pattern detected from {packet['source']}")
    
    return packet

def get_traffic_history():
    return active_streams
